brave/src/main/java/brave/secondary_sampling/SecondarySampling.java

```diff
@@ -184,12 +208,21 @@ public SecondarySampling build() {
     @Override protected Extra create(Extra parent) {
       return new Extra(parent);
     }
+
+    Extra create(Map<SecondarySamplingState, Boolean> initial) {
+      return initial.isEmpty() ? new Extra() : new Extra(initial);
+    }
   }
 
   static final class Extra extends MapPropagationFields<SecondarySamplingState, Boolean> {
     Extra() {
     }
 
+    // avoids copy-on-write for each keys on initial construction
+    Extra(Map<SecondarySamplingState, Boolean> initial) {
```

Comment:
TODO: we need benchmarks as this whole codebase can add a lot of overhead if we aren't careful.

Reply:
https://github.com/openzipkin-contrib/zipkin-secondary-sampling/issues/29